### PR TITLE
CT: Refactor Caution Flag #6854

### PR DESCRIPTION
**Pull request by zurbergram**
## Description
https://github.com/department-of-veterans-affairs/va.gov-team/issues/6854

Database migration: https://github.com/department-of-veterans-affairs/gibct-data-service/pull/611

## Testing done
- upload hcm CSV and mou CSV
- generate a preview
- publish preview
- Do either `GET /v0/institutions?name=STEVENS-HENAGER` or `GET /v0/institutions/31940744` and look for `caution_flags` attribute

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by elder-joseph — APPROVED** [review]
No issues here, it makes sense to add the array to `institutions`

File: spec/models/institution_builder_spec.rb [hunk 1]
```diff
@@ -560,73 +564,34 @@ def check_ipeds_ic_ay
...
-          expect(institutions.first.caution_flag_reason).to eq(reason)
+          expect(CautionFlag
+                     .where({ source: CautionFlag::SOURCES[:sec_702],
+                              version_id: Version.current_preview.id })
+                     .count).to be > 0
         end
 
-        it 'prefers Sec702School over Section702' do
-          sec702_school_institution = create :weam, :institution_builder, :private
+        it 'has flags when prefers Sec702School over Section702' do
```
Comment (shawkey-daniel) [inline spec/models/institution_builder_spec.rb]: Should this be no flags?

File: spec/models/institution_builder_spec.rb [hunk 2]
```diff
@@ -560,73 +564,34 @@ def check_ipeds_ic_ay
...
-
-        it 'is left unaltered when sec_702 is true' do
-          create :accreditation_institute_campus
-          create :accreditation_action_probationary
-          create :sec702_school, :institution_builder, sec_702: true
-          described_class.run(user)
-          expect(institutions.first.caution_flag_reason).to match(/Probation or Equivalent/)
-          expect(institutions.first.caution_flag_reason).not_to match(/Tuition/)
+          expect(CautionFlag
+                     .where({ source: CautionFlag::SOURCES[:mou],
```
Comment (shawkey-daniel) [inline spec/models/institution_builder_spec.rb]: Should this be 702?

**Review by shawkey-daniel — CHANGES_REQUESTED** [review]

File: spec/models/institution_builder_spec.rb (hunk 1 above)
Comment (zurbergram) [reply spec/models/institution_builder_spec.rb]: fixed

File: spec/models/institution_builder_spec.rb (hunk 2 above)
Comment (zurbergram) [reply spec/models/institution_builder_spec.rb]: fixed

**Review by shawkey-daniel — APPROVED** [review]
LGTM

**Review by sonntag-adam — APPROVED** [review]

File: app/models/institution.rb [hunk 3]
```diff
@@ -156,6 +156,7 @@ class Institution < ApplicationRecord
   }.freeze
 
   has_many :yellow_ribbon_programs, dependent: :destroy
+  has_many :caution_flags, dependent: :destroy, inverse_of: :institution
   has_many :institution_programs, -> { order(:description) }, inverse_of: :institution, dependent: :nullify
   has_many :versioned_school_certifying_officials, -> { order 'priority, last_name' }, inverse_of: :institution
```
Comment (johnpaulashenfelter) [inline app/models/institution.rb]: Might be easier to read in alphabetical order

**Review by johnpaulashenfelter — APPROVED** [review]
👍 approving, but it has failures (that are ignorable) on CodeClimate. @zurbergram @shawkey-daniel I believe you need either write or admin permission on the GitHub repo to be able to ignore the CodeClimate violations. I have that permissions on vets-api and can ignore this kind of issue but I cannot here. I *believe* any of your devs can ignore things like that if they have the right permissions.

File: app/models/institution.rb (hunk 3 above)
Comment (zurbergram) [reply app/models/institution.rb]: reordered

**Pull request merged**